File: solutions/1074. Number of Submatrices That Sum to Target/1074.cpp
// Time complexity: O(mn^2)
// Space complexity: O(m)
class Solution {
 public:
  int numSubmatrixSumTarget(const vector<vector<int>>& matrix, int target) {
    const int m = matrix.size();
    const int n = matrix[0].size();
    int ans = 0;

    for (auto& row : matrix)
      for (int i = 1; i < n; ++i)
        row[i] += row[i - 1];

    for (int baseCol = 0; baseCol < n; ++baseCol)
      for (int j = baseCol; j < n; ++j) {
        unordered_map<int, int> prefixCount{{0, 1}};
        int sum = 0;
        for (int i = 0; i < m; ++i) {
          if (baseCol > 0)
            sum -= matrix[i][baseCol - 1];
          sum += matrix[i][j];
          if (const auto it = prefixCount.find(sum - target);
              it != prefixCount.cend())
            ans += it->second;
          ++prefixCount[sum];
        }
      }

    return ans;
  }
};

File: solutions/1074. Number of Submatrices That Sum to Target/1074.rs
// Time complexity: O(mn^2)
// Space complexity: O(m)
use std::collections::HashMap;

impl Solution {
    pub fn num_submatrix_sum_target(matrix: Vec<Vec<i32>>, target: i32) -> i32 {
        let m = matrix.len();
        let n = matrix[0].len();
        let mut matrix = matrix;
        let mut ans = 0;

        for row in matrix.iter_mut() {
            for i in 1..n {
                row[i] += row[i - 1];
            }
        }

        for base_col in 0..n {
            for j in base_col..n {
                let mut prefix_count = HashMap::new();
                prefix_count.insert(0, 1);
                let mut sum = 0;
                for i in 0..m {
                    if base_col > 0 {
                        sum -= matrix[i][base_col - 1];
                    }
                    sum += matrix[i][j];
                    if let Some(&count) = prefix_count.get(&(sum - target)) {
                        ans += count;
                    }
                    *prefix_count.entry(sum).or_insert(0) += 1;
                }
            }
        }

        ans
    }
}
